pkg/storage/stores/shipper/uploads/table.go

```diff
@@ -0,0 +1,321 @@
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"go.etcd.io/bbolt"
+)
+
+const (
+	// create a new db sharded by time based on when write request is received
+	shardDBsByDuration = 15 * time.Minute
+
+	// retain dbs for specified duration after they are modified to avoid keeping them locally forever.
+	// this period should be big enough than shardDBsByDuration to avoid any conflicts
+	dbRetainPeriod = time.Hour
+)
+
+type BoltDBIndexClient interface {
+	QueryDB(ctx context.Context, db *bbolt.DB, query chunk.IndexQuery, callback func(chunk.IndexQuery, chunk.ReadBatch) (shouldContinue bool)) error
+	WriteToDB(ctx context.Context, db *bbolt.DB, writes local.TableWrites) error
+}
+
+type StorageClient interface {
+	PutObject(ctx context.Context, objectKey string, object io.ReadSeeker) error
+}
+
+// Table is a collection of multiple dbs created for a same table by the ingester.
+// All the public methods are concurrency safe and take care of mutexes to avoid any data race.
+type Table struct {
+	name              string
+	path              string
+	uploader          string
+	storageClient     StorageClient
+	boltdbIndexClient BoltDBIndexClient
+
+	dbs    map[string]*bbolt.DB
+	dbsMtx sync.RWMutex
+
+	uploadedDBsMtime    map[string]time.Time
+	uploadedDBsMtimeMtx sync.RWMutex
+}
+
+// NewTable create a new Table without looking for any existing local dbs belonging to the table.
+func NewTable(path, uploader string, storageClient StorageClient, boltdbIndexClient BoltDBIndexClient) (*Table, error) {
+	err := chunk_util.EnsureDirectory(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return newTableWithDBs(map[string]*bbolt.DB{}, path, uploader, storageClient, boltdbIndexClient)
+}
+
+// LoadTable loads local dbs belonging to the table and creates a new Table with references to dbs if there are any otherwise it doesn't create a table
+func LoadTable(path, uploader string, storageClient StorageClient, boltdbIndexClient BoltDBIndexClient) (*Table, error) {
+	dbs, err := loadBoltDBsFromDir(path)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(dbs) == 0 {
+		return nil, nil
+	}
+
+	return newTableWithDBs(dbs, path, uploader, storageClient, boltdbIndexClient)
+}
+
+func newTableWithDBs(dbs map[string]*bbolt.DB, path, uploader string, storageClient StorageClient, boltdbIndexClient BoltDBIndexClient) (*Table, error) {
+	return &Table{
+		name:              filepath.Base(path),
+		path:              path,
+		uploader:          uploader,
+		storageClient:     storageClient,
+		boltdbIndexClient: boltdbIndexClient,
+		dbs:               dbs,
+		uploadedDBsMtime:  map[string]time.Time{},
+	}, nil
+}
+
+// Query serves the index by querying all the open dbs.
+func (lt *Table) Query(ctx context.Context, query chunk.IndexQuery, callback chunk_util.Callback) error {
+	lt.dbsMtx.RLock()
+	defer lt.dbsMtx.RUnlock()
+
+	for _, db := range lt.dbs {
+		if err := lt.boltdbIndexClient.QueryDB(ctx, db, query, callback); err != nil {
```

Comment:
I wonder if this shouldn't be run on multiple goroutine ? Technically reading a file is not CPU bound so you should gain something by not waiting an operation and fire more at once.

Reply:
I'm not sure but a benchmark could help up figure this out.

Reply:
I think let us have a basic version out, we can do all these optimizations when we see a problem there. We already parallelize queries for different tables, we can do the same here with bounded concurrency. What do you think?

Reply:
Yep no problem !